workspaces/ui-v2/src/pages/docs/components/ShapeEditor.tsx

```diff
@@ -58,34 +85,89 @@ export const ShapeEditor: FC<{
   );
 };
 
+const editableTypes: Set<JsonType> = new Set([
+  JsonType.UNDEFINED,
+  JsonType.NULL,
+]);
+const getInitialTypes = (fields: IFieldDetails): Set<JsonType> => {
+  const initialTypes: Set<JsonType> = new Set();
+  for (const type of editableTypes) {
+    if (type === JsonType.UNDEFINED && !fields.required) {
+      initialTypes.add(type);
+    } else if (fields.shapes.find((shape) => shape.jsonType === type)) {
+      initialTypes.add(type);
+    }
+  }
+  return initialTypes;
+};
+
 const Row: FC<{
   field: IFieldDetails;
-  description: string;
   selected: boolean;
 
-  onChangeDescription?: (fieldId: string, description: string) => void;
+  isFieldRemoved?: ComponentProps<typeof ShapeEditor>['isFieldRemoved'];
+  onToggleRemove?: ComponentProps<typeof ShapeEditor>['onToggleRemove'];
+  onChangeDescription?: ComponentProps<
+    typeof ShapeEditor
+  >['onChangeDescription'];
+  onChangeFieldType?: ComponentProps<typeof ShapeEditor>['onChangeFieldType'];
   onSelect?: () => void;
 }> = function ShapeEditorRow({
   field,
-  description,
   selected,
   onChangeDescription,
+  onChangeFieldType,
   onSelect,
+  isFieldRemoved,
+  onToggleRemove,
 }) {
   const classes = useStyles();
+  const initialDescription = field.contribution['value'];
+  const initialFieldTypes = getInitialTypes(field);
```

Comment:
We need the initial description and field types to compare to the current input (such that we know if we need to reset and remove state) - this is the simplest way to achieve this and isn't a huge amount of extra work